packages/types-terra/src/interfaces.ts

```diff
@@ -26,5 +26,23 @@ export interface TerraBlock {
 
 export interface TerraEvent {
   event: EventsByType;
-  block: BlockInfo;
+  blockNumber: number;
+  blockHash: string;
+  blockTimestamp: string;
+  msgIndex: number;
+  txInfo: TxInfo;
+  log: TxLog;
+  transactionHash: string;
+}
+
+export interface TerraCall {
+  from: string;
+  to: string;
+  fee: Fee.Data;
+  data: MsgExecuteContract.Data;
+  hash: string;
+  blockNumber: number;
+  blockHash: string;
+  timestamp: string;
+  signatures: string[];
```

Comment:
I think we can improve these types to be more like Substrate

```ts

interface TerraBlock {
  block: BlockInfo;
  txs: TxInfo[];
}

interface TerraCall {
  idx: number;
  block: TerraBlock;
  tx: TxInfo;
}

interface TerraEvent {
  idx: number; // To avoid doing https://github.com/DeveloperInProgress/terra-contract-starter/blob/main/src/mappings/mappingHandlers.ts#L17
  block: TerraBlock;
  call: TerraCall;
  event: TxLog; // Not sure this is the type we want
}

```

Reply:
we could msg_index from logs as idx for TerraEvent. But, what shall we assign for idx in TerraCall?

Reply:
Here is what I suggest:

```
export interface TerraBlock {
  block: BlockInfo;
  txs: TxInfo[];
}

export interface TerraEvent {
  event: EventsByType;
  block: TerraBlock;
  msgIndex: number;
  txInfo: TxInfo;
  log: TxLog;
}

export interface TerraCall {
  idx: number;
  data: MsgExecuteContract.Data;
  block: TerraBlock;
  tx: TxInfo;
}
```

In TerraEvent, msgIndex could be used along transaction hash to create an unique primary key. Also, provide the event data as EventsByType for easy access and also the corresponding TxLog for those who would like to deal with the raw log.

In TerraCall, should idx be a random generated number?

Reply:
By the looks of things we want another type of handler for `Messages`. Because `Blocks` have `Transactions` which have `Messages` which  have `Events`. We're currently skipping over the extra layer of depth.

We should make the following changes:

- Rename `Call` to `Transaction` to be consistent with Terra naming conventions
- Introduce `Message` handler type.
    - Filter can have a `type` which should match to these [types](https://github.com/terra-money/terra.js/blob/9e5f553/src/core/Msg.ts#L239)
    - Filter can also match any key value on the [Amino type](https://github.com/terra-money/terra.js/blob/9e5f553de3ff3e975eaaf91b1f06e45658b1a5e0/src/core/wasm/msgs/MsgExecuteContract.ts#L113) e.g. `- sender: terra1j66jatn3k50hjtg2xemnjm8s7y8dws9xqa5y8w`. We can keep validation to simple object validation for now
- `Event` filters can keep the `type` filter as well as adding Message filters. 
   ```yaml
    - type: transfer
    - messageFilter:
        - type: '/terra.wasm.v1beta1.MsgExecuteContract'
        - contract: terra1627ldjvxatt54ydd3ns6xaxtd68a2vtyu7kakj
   ```

Let me know if this doesn't make sense and I can explain further.

> In TerraCall, should idx be a random generated number?

We don't want to randomly generate numbers

> Also, provide the event data as EventsByType

I don't think this is very helpful except for in a block handler.

Reply:
Agree, the idx should be the ordered index of the event/call on the block if possible